text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+
+Another challenge is that some features such as App Mesh require resources and settings
+across multiple construct types. For example App Mesh requires that the service be
+modified to have a CloudMap namespace, it must modify the task definition to have an
+Envoy sidecar and proxy settings, and it must have created App Mesh specific resources
+like Virtual Node and Virtual Service. There is no single place for all of this cross
+construct logic to live in the current construct architecture.
+
+Last but not least the existing approach to L3 constructs has an `Ec2` and a `Fargate`
+version of each L3 construct. Ideally there would be a way for customers to create a
+service which is agnostic to whether it is deployed on EC2 or Fargate.
+
+## Tenants
+
+To solve the problems with ECS construct implemention we propose adding a new `Service`
+class with the following goals:
+
+- __One service class to learn__: Instead of a collection of many different L3 classes
+  that all have different names there is a single class which can implement any number
+  of different forms depending on what the customer wants. Customers don't need to learn
+  the names for different L3 constructs. All ECS architectures start the same way: with
+  the same `Service` class.
+- __Easy to extend and customize__: It is easy to extend the `Service` class with new
+  behaviors and features, in any combination required. The `Service` class comes with
+  batteries included: a set of easy to use extensions created and maintained by the AWS team.
+  However there are also clear instructions and examples of how customers can create their
+  own custom extensions. Third parties will be enabled to also create and distribute
+  extensions for their own services, for example a provider like Datadog could easily
+  build an extension that automatically adds Datadog agent to an ECS service.
+- __Extensions self configure and self provision__: Service extensions provision their
+  own resources as well as configuring the ECS service to use those resources. For example
+  a load balancer extension creates its own load balancer and also attaches the ECS service
+  to the load balancer. An App Mesh extension creates its own App Mesh virtual service and
+  virtual node as well as configuring the ECS service and ECS task definition to have the
+  right Envoy sidecar and proxy settings.
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a
+  service extension. Extensions know of each other's existence. For example the application's
+  container is aware that there is an Envoy proxy and that it must wait until the Envoy
+  proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing the capacity strategy
+  in the class name customers just create a service, and add it to an environment. The
+  service automatically modifies its settings as needed to adjust to the capacity available
+  in the environment.
```

Comment:
Are there certain extensions that would need to take into account restrictions in capacity/orchestration providers?

Reply:
Yes, it is likely in the future that an extension could choose to deploy itself as either a sidecar or as a cluster daemon depending on whether the extension is being added to a Fargate or EC2 environment